### PR TITLE
Pytorch RT-BENE

### PR DESCRIPTION
**Pull request by ahmed-alhindawi**
Hello,

I've pytorched RT-BENE for easier experimentation.
Let me know what you think of this pull request.

### CODE REVIEW

File: rt_bene_model_training/README.md [hunk 1]
```diff
@@ -31,7 +31,7 @@ For pip users: `pip install tensorflow-gpu numpy tqdm opencv-python scikit-learn
 This code was used to train the blink estimator for RT-BENE. The labels for the RT-BENE blink dataset are contained in the [rt_bene_dataset](../rt_bene_dataset) directory. The images corresponding to the labels can be downloaded from the RT-GENE dataset (labels are only available for the "noglasses" part): [download](https://zenodo.org/record/2529036) [(alternative link)](https://goo.gl/tfUaDm). Please run `python train_blink_model.py --help` to see the required arguments to train the model.
 
 ## Model testing code
-Evaluation code for a 3-fold evaluation is provided in the [evaluate_blink_model.py](./evaluate_blink_model.py) file. An example to train and evaluate an ensemble of models can be found in [train_and_evaluate.py](./train_and_evaluate.py). Please run `python train_and_evaluate.py --help` to see the required arguments.
+Evaluation code for a 3-fold evaluation is provided in the [evaluate_blink_model.py](tensorflow/evaluate_blink_model.py) file. An example to train and evaluate an ensemble of models can be found in [train_and_evaluate.py](tensorflow/train_and_evaluate.py). Please run `python train_and_evaluate.py --help` to see the required arguments.
```
Comment (Tobias-Fischer) [inline rt_bene_model_training/README.md:34]: Could you please add a section in the readme referring to the pytorch code?

File: rt_bene_model_training/pytorch/rtbene_dataset.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            self._transform = transforms.Compose([transforms.Resize((224, 224), Image.BICUBIC),
+                                                  transforms.ToTensor(),
+                                                  transforms.Normalize(mean=[0.485, 0.456, 0.406],
+                                                                       std=[0.229, 0.224, 0.225])])
+
+        _wanted_subjects = ["s{:03d}".format(_i) for _i in subject_list]
+
+        for grp_s_n in tqdm(_wanted_subjects, desc="Loading ({}) subject metadata...".format(loader_desc)):  # subjects
+            for grp_i_n, grp_i in h5_file[grp_s_n].items():  # images
+                if "image" in grp_i.keys() and "label" in grp_i.keys():
```
Comment (Tobias-Fischer) [inline rt_bene_model_training/pytorch/rtbene_dataset.py:31]: Is there ever a case where "image" and "label" are not in the keys?

File: rt_bene_model_training/pytorch/rtbene_dataset.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+                    image_dataset = grp_i["image"]
+                    label = grp_i["label"]
+                    if label == 1.0:
```
Comment (Tobias-Fischer) [inline rt_bene_model_training/pytorch/rtbene_dataset.py:34]: `self._positive_labels` and `self._total_labels` are set but never used

File: rt_bene_model_training/pytorch/util/GenerateRTBENEH5Dataset.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+from PIL import Image, ImageFilter, ImageOps
+from torchvision import transforms
+from tqdm import tqdm
+
+script_path = os.path.dirname(os.path.realpath(__file__))
+
+# Augmentations following `prepare_dataset.m`: randomly crop and resize the image 10 times,
+# along side two blurring stages, grayscaling and histogram normalisation
+_required_size = (224, 224)
+_transforms_list = [transforms.RandomResizedCrop(size=_required_size, scale=(0.85, 1.0)),  # equivalent to random 5px from each edge
```
Comment (Tobias-Fischer) [inline rt_bene_model_training/pytorch/util/GenerateRTBENEH5Dataset.py:17]: Is that really what's happening?!? Can't remember doing that, looks a little crazy .. how is that different from doing a single `RandomResizedCrop` with different scale factor?

File: rt_bene_model_training/pytorch/util/GenerateRTBENEH5Dataset.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    augmented_images.append(np.array(image))
+
+    return np.array(augmented_images, dtype=np.uint8)
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description='Estimate gaze from images')
+    parser.add_argument('--rt_bene_root', type=str, required=True, nargs='?', help='Path to the base directory of RT_GENE')
+    parser.add_argument('--augment_dataset', type=bool, required=False, default=False, help="Whether to augment the dataset with predefined transforms")
+    parser.add_argument('--compress', action='store_true', dest="compress")
```
Comment (Tobias-Fischer) [inline rt_bene_model_training/pytorch/util/GenerateRTBENEH5Dataset.py:45]: Could you please add a help to that?

**Review by Tobias-Fischer — CHANGES_REQUESTED** [review]
Hi @ahmed-alhindawi,
I added some comments inline. Bigger question though - have you tried/tested this? Are the results similar to the TF version? As far as I can see there is no code to do the 3-fold evaluation, which is not a big deal, but it would be good to know if the models that are trained are sensible.

File: rt_bene_model_training/pytorch/util/GenerateRTBENEH5Dataset.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    parser.add_argument('--no-compress', action='store_false', dest="compress")
+    parser.set_defaults(compress=False)
+    args = parser.parse_args()
+
+    _compression = "lzf" if args.compress is True else None
+
+    subject_path = [os.path.join(args.rt_bene_root, "s{:03d}_noglasses/".format(_i)) for _i in range(0, 17)]
+
+    hdf_file = h5py.File(os.path.abspath(os.path.join(args.rt_bene_root, "rtbene_dataset.hdf5")), mode='w')
```
Comment (Tobias-Fischer) [inline rt_bene_model_training/pytorch/util/GenerateRTBENEH5Dataset.py:54]: Use a `with` statement here

**Comment by ahmed-alhindawi** [comment]
> I added some comments inline. Bigger question though - have you tried/tested this? Are the results similar to the TF version? As far as I can see there is no code to do the 3-fold evaluation, which is not a big deal, but it would be good to know if the models that are trained are sensible.

Hey, yes it works and trains nicely. You're correct, there isn't any evaluation code yet; that's something to do this week but the fundamentals to perform 3-fold validation is there; the Trainer trains on specific train/validation subjects who are the same splits as the TF version.

File: rt_bene_model_training/pytorch/util/GenerateRTBENEH5Dataset.py (hunk 4 above)
Comment (ahmed-alhindawi) [reply rt_bene_model_training/pytorch/util/GenerateRTBENEH5Dataset.py:17]: I'm sorry I don't understand your question...
To be fair, that's somewhat deviating from the original TF code; the scale of (0.85 to 1.0) is a randomized scaling factor as a form of data augmentation...

File: rt_bene_model_training/pytorch/rtbene_dataset.py (hunk 2 above)
Comment (ahmed-alhindawi) [reply rt_bene_model_training/pytorch/rtbene_dataset.py:31]: not in RT-BENE dataset, similar to the training code for RT-GENE, I was hoping to future-proof it if there was ever another dataset out there that we can use.

File: rt_bene_model_training/pytorch/util/GenerateRTBENEH5Dataset.py (hunk 6 above)
Comment (ahmed-alhindawi) [reply rt_bene_model_training/pytorch/util/GenerateRTBENEH5Dataset.py:54]: Nice!

File: rt_bene_model_training/pytorch/util/GenerateRTBENEH5Dataset.py (hunk 5 above)
Comment (ahmed-alhindawi) [reply rt_bene_model_training/pytorch/util/GenerateRTBENEH5Dataset.py:45]: Done!

File: rt_bene_model_training/pytorch/rtbene_dataset.py (hunk 2 above)
Comment (Tobias-Fischer) [reply rt_bene_model_training/pytorch/rtbene_dataset.py:29]: Gotcha

File: rt_bene_model_training/README.md (hunk 1 above)
Comment (ahmed-alhindawi) [reply rt_bene_model_training/README.md:43]: Done.

File: rt_bene_model_training/README.md [hunk 7]
```diff
@@ -25,13 +25,24 @@ RT-BENE was supported by the EU Horizon 2020 Project PAL (643783-RIA) and a Roya
 More information can be found on the Personal Robotic Lab's website: <https://www.imperial.ac.uk/personal-robotics/software/>.
 
 ## Requirements
+### Tensorflow
 For pip users: `pip install tensorflow-gpu numpy tqdm opencv-python scikit-learn` or for conda users: `conda install tensorflow-gpu numpy tqdm opencv scikit-learn`
+### Pytorch
+
```
Comment (Tobias-Fischer) [inline rt_bene_model_training/README.md:31]:
```suggestion
For conda users: conda install -c conda-forge numpy scipy tqdm pillow rospkg opencv scikit-learn h5py matplotlib pytorch-lightning && conda install -c pytorch pytorch torchvision
```

File: rt_bene_model_training/pytorch/evaluate_model.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+from rt_bene_model_training.pytorch.rtbene_dataset import RTBENEH5Dataset
+
+if __name__ == "__main__":
+    root_dir = os.path.dirname(os.path.realpath(__file__))
+    _root_parser = ArgumentParser(add_help=False)
+    _root_parser.add_argument('--hdf5_file', type=str,
+                              default=os.path.abspath(os.path.join(root_dir, "../../RT_BENE/rtbene_dataset.hdf5")))
+    _root_parser.add_argument('--dataset', type=str, choices=["rt_bene"], default="rt_bene")
+    _root_parser.add_argument('--model_net_dir', type=str, required=True)
+    _root_parser.add_argument('--seed', type=int, default=0)
```
Comment (Tobias-Fischer) [inline rt_bene_model_training/pytorch/evaluate_model.py:22]:
```suggestion
```

File: rt_bene_model_training/README.md [hunk 9]
```diff
@@ -25,13 +25,24 @@ RT-BENE was supported by the EU Horizon 2020 Project PAL (643783-RIA) and a Roya
...
 
 ## Model testing code
-Evaluation code for a 3-fold evaluation is provided in the [evaluate_blink_model.py](./evaluate_blink_model.py) file. An example to train and evaluate an ensemble of models can be found in [train_and_evaluate.py](./train_and_evaluate.py). Please run `python train_and_evaluate.py --help` to see the required arguments.
+### Tensorflow
+Evaluation code for a 3-fold evaluation is provided in the [evaluate_blink_model.py](tensorflow/evaluate_blink_model.py) file. An example to train and evaluate an ensemble of models can be found in [train_and_evaluate.py](tensorflow/train_and_evaluate.py). Please run `python train_and_evaluate.py --help` to see the required arguments.
 
 ![Results](../assets/rt_bene_precision_recall.png)
+### Pytorch
+No evaluation code is available yet, this is to be added.
```
Comment (Tobias-Fischer) [inline rt_bene_model_training/README.md:47]: Looks like there is now?

File: rt_bene_model_training/pytorch/train_model.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+        if _hyperparams.k_fold_validation:
+            # 6 is discarded
+            _train_subjects.append([1, 2, 8, 10])
+            _train_subjects.append([3, 4, 7, 9])
+            _train_subjects.append([5, 12, 13, 14])
+
+            _valid_subjects.append([0, 11, 15, 16])
+            _valid_subjects.append([0, 11, 15, 16])
+            _valid_subjects.append([0, 11, 15, 16])
+        else:  # we want to train with the entire dataset
```
Comment (Tobias-Fischer) [inline rt_bene_model_training/pytorch/train_model.py:158]:
```suggestion
        else:  # we want to train with the entire dataset
            print('Training on the whole dataset - do not use the trained model for evaluation purposes!')
```

File: rt_bene_model_training/pytorch/train_model.py [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+            _train_subjects.append(
+                [0, 1, 2, 3, 4, 5, 8, 9, 10, 11, 12, 13, 14, 15, 16])  # 6 is discarded, 7 is used for validation
+            _valid_subjects.append([7])  # Note that this is a hack and should not be used to get results for papers
```
Comment (Tobias-Fischer) [inline rt_bene_model_training/pytorch/train_model.py:161]:
```suggestion
            _valid_subjects.append([7])
```

**Review by Tobias-Fischer — CHANGES_REQUESTED** [review]
Looking good now! I made a couple of suggestions - what do you think?

File: rt_bene_model_training/README.md (hunk 9 above)
Comment (ahmed-alhindawi) [reply rt_bene_model_training/README.md:47]: There is indeed! I'm doing a run now and will upload the ROC and PRC curves to see how they match. Prelimiary results are that theyre identical but gimme 12 hours...

Comment (Tobias-Fischer) [reply rt_bene_model_training/README.md:47]: Great, thanks :)

Comment (ahmed-alhindawi) [reply rt_bene_model_training/README.md:47]: Thanks @Tobias-Fischer - I've uploaded a graph of the ROC and PRC curves in the assets - it's at least non-inferior to the tensorflow version.
Let me know what you think.

File: rt_bene_model_training/pytorch/train_model.py [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+                                              save_top_k=3,
+                                              filename=f'fold={fold}-' + '{epoch}-{val_loss:.2f}')
+
+        # start training
+        trainer = Trainer(gpus=_hyperparams.gpu,
+                          callbacks=[checkpoint_callback],
+                          precision=32,
+                          progress_bar_refresh_rate=1,
+                          min_epochs=5 if _hyperparams.augment else 3,
+                          max_epochs=20 if _hyperparams.augment else 5,
```
Comment (Tobias-Fischer) [inline rt_bene_model_training/pytorch/train_model.py:187]: Shouldn't there be fewer epochs in we augment?

File: rt_bene_model_training/pytorch/util/GenerateRTBENEH5Dataset.py [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+    subject_path = [os.path.join(args.rt_bene_root, "s{:03d}_noglasses/".format(_i)) for _i in range(0, 17)]
+
+    with h5py.File(os.path.abspath(os.path.join(args.rt_bene_root, "rtbene_dataset.hdf5")), mode='w') as hdf_file:
+        for subject_id, subject_data in enumerate(subject_path):
+            subject_id = str("s{:03d}".format(subject_id))
+            subject_grp = hdf_file.create_group(subject_id)
+            with open(os.path.join(args.rt_bene_root, "{}_blink_labels.csv".format(subject_id)), "r") as f:
+                _lines = f.readlines()
+
+                for line in tqdm(_lines, desc="Subject {}".format(subject_id)):
```
Comment (Tobias-Fischer) [inline rt_bene_model_training/pytorch/util/GenerateRTBENEH5Dataset.py:37]: I think this can be un-indented.

**Comment by Tobias-Fischer** [comment]
Two tiny comments left, well done!

File: rt_bene_model_training/pytorch/util/GenerateRTBENEH5Dataset.py (hunk 13 above)
Comment (ahmed-alhindawi) [reply rt_bene_model_training/pytorch/util/GenerateRTBENEH5Dataset.py:37]: ah sure, it's a stylistic thing I guess but I suppose the for loop isn't dependent on reading the file....Fair enough!

File: rt_bene_model_training/pytorch/train_model.py (hunk 12 above)
Comment (ahmed-alhindawi) [reply rt_bene_model_training/pytorch/train_model.py:187]: To be honest, the epochs length is rather arbitrary at the moment - this stems from if we dont augment in the Trainer but augment at data generation time to make them equal but I've been fiddling with the augment number and not augment number. 
I should make this an argument in the parser with some sensible defaults...

**Comment by Tobias-Fischer** [comment]
Great - ready to merge if you think it's ready.

**Comment by ahmed-alhindawi** [comment]
> Great - ready to merge if you think it's ready.

Two things:
1 - I realised there's no way to perform inference with the models, so let me make that code today
2 - I'm going to change the evaluate_mode.py to a jupyter notebook to facilitate more data exploration.

**Comment by ahmed-alhindawi** [comment]
Ready if you're happy?

File: rt_gene/launch/estimate_blink.launch [hunk 14]
```diff
@@ -2,10 +2,12 @@
     <arg name="threshold" default="0.425" />
     <arg name="viz_topic" default="/subjects/blink_images" />
     <arg name="viz" default="True" />
-    <arg name="device_id_blinkestimation"  default="/gpu:0" /> <!-- tensorflow format, e.g. /gpu:0 or /cpu:0 -->
+    <arg name="blink_backend" default="tensorflow" />
```
Comment (Tobias-Fischer) [inline rt_gene/launch/estimate_blink.launch:5]:
```suggestion
    <arg name="blink_backend" default="tensorflow" /> <!-- can be tensorflow or pytorch -->
```

File: rt_gene/src/rt_bene/estimate_blink_pytorch.py [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+    def predict(self, left_eyes, right_eyes):
+        transformed_left = torch.stack(left_eyes).to(self.device_id)
+        transformed_right = torch.stack(right_eyes).to(self.device_id)
+
+        result = [torch.sigmoid(model(transformed_left, transformed_right)).detach().cpu() for model in self._models]
+        result = torch.stack(result, dim=1)
+        result = torch.mean(result, dim=1).numpy()
+        return result
+
+    def inputs_from_images(self, left, right):
```
Comment (Tobias-Fischer) [inline rt_gene/src/rt_bene/estimate_blink_pytorch.py:47]: No flipping needed here? Elsewhere you flip the images, don't you?

**Review by Tobias-Fischer — CHANGES_REQUESTED** [review]

File: rt_gene/src/rt_bene/estimate_blink_pytorch.py (hunk 15 above)
Comment (ahmed-alhindawi) [reply rt_gene/src/rt_bene/estimate_blink_pytorch.py:47]: The tensorflow training requires the right image to be flipped; but I didn't see a particular reason for it so the dataset generation, training and inference do not flip in the pytorch version...
I could do it to maintain consistency between the two versions but I personally feel that we don't need it.

Comment (Tobias-Fischer) [reply rt_gene/src/rt_bene/estimate_blink_pytorch.py:47]: Sounds good to me, thanks for clarifying :)

Comment (Tobias-Fischer) [reply rt_gene/src/rt_bene/estimate_blink_pytorch.py:47]: By the way, flipping was done originally for weight sharing - I can't remember whether we ended up using it. In this case the idea is that the two images are more similar if they're flipped.

**Comment by Tobias-Fischer** [comment]
Many thanks again @ahmed-alhindawi! Nice job :)

File: rt_gene/src/rt_bene/estimate_blink_pytorch.py (hunk 15 above)
Comment (ahmed-alhindawi) [reply rt_gene/src/rt_bene/estimate_blink_pytorch.py:47]: Ah that makes sense - in the TF version, there isn't any particular weight sharing going on so I don't think it is being used. There's two independent neural networks that do inference on each eye...much like RT-GENE.
